cachito/web/models.py

```diff
@@ -186,6 +186,30 @@ def to_purl(self):
             # see https://github.com/package-url/purl-spec/issues/63 for further reference
             purl_name = self.name.replace("/", "%2F")
             return f"pkg:golang/{purl_name}@{self.version}"
+
+        elif self.type == "npm":
+            purl_name = self.name.replace("@", "%40")
+            match = re.match(r"(?P<protocol>[^:]+):(?://)?(?P<suffix>.+)", self.version)
+            if not match:
+                return f"pkg:npm/{purl_name}@{self.version}"
+            protocol = match.group("protocol")
+            suffix = match.group("suffix")
+            if protocol in ("github", "gitlab", "bitbucket"):  # github:namespace/name#ref
+                match_forge = re.match(
+                    r"(?P<namespace>[^/]+)/(?P<name>[^#]+)#(?P<version>.+)$", suffix
+                )
+                if not match_forge:
+                    raise ValueError(f"Could not convert version {self.version} to purl")
+                forge = match_forge.groupdict()
+                return f"pkg:{protocol}/{forge['namespace']}/{forge['name']}@{forge['version']}"
+            elif protocol in ("git", "git+http", "git+https", "git+ssh"):
+                return f"pkg:npm/{purl_name}?vcs_url={self.version}"
+            elif protocol in ("http", "https"):
+                # assume this is an npm repository
+                return f"pkg:npm/{purl_name}?repository_url={self.version}"
```

Comment:
I don't think `repository_url` is quite right in this case as I think it would only apply to if the dependency came from an alternate npm registry.